docs/decisions/0033-rag-in-sk.md

```diff
@@ -0,0 +1,297 @@
+
+Ideally, each connector should be always up-to-date and support new functionality. For some connectors maintenance cost is low, since there are no breaking changes included in new features or vector DB provides .NET SDK which is relatively easy to re-use. For other connectors maintenance cost is high, since some of them are still in `alpha` or `beta` development stage, breaking changes can be included or .NET SDK is not provided, which makes it harder to update.
+
+### IMemoryStore interface
+
+Each memory connector implements `IMemoryStore` interface with methods like `CreateCollectionAsync`, `GetNearestMatchesAsync` etc., so it can be used as part of `TextMemoryPlugin`.
+
+By implementing the same interface, each integration is aligned, which makes it possible to use different vector DBs at runtime. At the same time it is disadvantage, because each vector DB can work differently, and it becomes harder to fit all integrations into already existing abstraction. For example, method `CreateCollectionAsync` from `IMemoryStore` is used when application tries to add new record to vector DB to the collection, which doesn't exist, so before insert operation, it creates new collection. In case of [Pinecone](https://www.pinecone.io/) vector DB, this scenario is not supported, because Pinecone index creation is an asynchronous process - API service will return 201 Created HTTP response with following property in response body (index is not ready for usage):
+
+```json
+{
+    // Other properties...
+    "status": {
+        "ready": false,
+        "state": "Initializing"
+    }
+}
+```
+
+In this case, it's impossible to insert a record to database immediately, so HTTP polling or similar mechanism should be implemented to cover this scenario.
+
+### MemoryRecord as storage schema
+
+`IMemoryStore` interface uses `MemoryRecord` class as storage schema in vector DB. This means that `MemoryRecord` properties should be aligned to all possible connectors. As soon as developers will use this schema in their databases, any changes to schema may break the application, which is not flexible approach.
+
+`MemoryRecord` contains property `ReadOnlyMemory<float> Embedding` for embeddings and `MemoryRecordMetadata Metadata` for embeddings metadata. `MemoryRecordMetadata` contains properties like:
+
+- `string Id` - unique identifier.
+- `string Text` - data-related text.
+- `string Description` - optional title describing the content.
+- `string AdditionalMetadata` - field for saving custom metadata with a record.
+
+Since `MemoryRecord` and `MemoryRecordMetadata` are not sealed classes, it should be possible to extend them and add more properties as needed. Although, current approach still forces developers to have specific base schema in their vector DBs, which ideally should be avoided. Developers should have the ability to work with any schema of their choice, which will cover their business scenarios (similarly to Code First approach in Entity Framework).
```

Comment:
> If the developer extends MemoryRecord/MemoryRecordMetadata, I'm assuming it isn't actually expected that SK IMemoryStore implementations actually pick those changes up (via reflection), is it? At least, that's not what I saw some/most implementations do - they just hard-code the several known metadata fields, nothing more.

That's correct, additional mechanism should be added to take extended classes into account. Even current implementations don't take into account all fields from MemoryRecord/MemoryRecordMetadata. Ideally, developer should have control over it, and connector should just construct a request based on all properties that were passed to it.

> That might mean that the SK abstraction should also go in a more weakly-typed, JSON-oriented direction; this would allow users to have whatever data shapes they want in their metadata.

I think the ideal scenario would be to support strongly-typed schema and weakly-typed JSON-based approach to cover both cases. I will probably investigate in more detail which approach is used in each connector and what is supported by vector DB out-of-the-box, to have a better understanding.